### ISSUE TITLE
Add Siamese Contrastive Loss function

### ISSUE THREAD

**Issue by arcAman07** [FluxML/Flux.jl#1880]
Addition of New Contastive Loss function which is used as a primary metric to calculate loss in siamese networks. If it needs to be added, I would love to work on the PR

**Comment by DhairyaLGandhi** [comment]
I don't know if that is a commonly used enough metric to be added to Flux wholesale. Are there cases where it is required more generally? It should be easy enough to be added to an implementation of Siamese networks I think.

**Comment by arcAman07** [comment]
Yeah I was working on a lot of similarity related ML models with fashion items,etc. It's a pretty easy loss function to add ( not a difficult task ) so was just thinking rather than users creating their own, it would be nice to add it along with other loss functions.
So if it can be added, I can work on the PR and try to do it by today.

**Comment by CarloLucibello** [comment]
Is there precedence for it in other deep learning libraries?

**Comment by arcAman07** [comment]
It is implemented in tensorflow. Not specfically added in pytorch but in this repo : https://github.com/KevinMusgrave/pytorch-metric-learning/ , he has written these important loss functions which people use to build models over pytorch. So it could be a useful add to the library as ppl could use it by just inserting one parameter ( the margin value ).

**Comment by darsnack** [comment]
As a community, we haven't agreed on a good place to define losses other than Flux. This is a well known loss that's complex enough to provide for users. I think we should at least consider a PR.

**Comment by darsnack** [comment]
As a community, we haven't agreed upon a place to define losses other than Flux. This is a well known loss that is complex enough that we could provide it for users. I think we should at least consider a PR.

**Comment by arcAman07** [comment]
Cool will start working on the implementation of this loss function and send a PR as soon as I get done with it.

**Comment by CarloLucibello** [comment]
Please also add a reference to the paper in the docstring and make sure that the interface is consistent with the other losses we have in the `Losses` module. I'm not sure which loss exactly you want to implement: is it what is called the `TripletMarginLoss`
in [pytorch](https://pytorch.org/docs/stable/generated/torch.nn.TripletMarginLoss.html#torch.nn.TripletMarginLoss)?

**Comment by arcAman07** [comment]
Will take great care while implementing the loss function. It isn't exactly the TripletMarginLoss function but something similar to it. 
Reference:  https://pyimagesearch.com/2021/01/18/contrastive-loss-for-siamese-networks-with-keras-and-tensorflow/

It has been implemented in keras too , will check in pytorch too.

**Comment by CarloLucibello** [comment]
A blog post is not an acceptable reference. 

So if we are talking about the `ContrastiveLoss` in [KevinMusgrave's repo](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/CONTENTS.md) the paper they refer to is [this one](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf)

> It has been implemented in keras too 

Can you post a link?

**Comment by arcAman07** [comment]
My apologies, i just shared the blog for the understanding of the loss function and how is it different from the TripletMarginLoss function. The paper you referred is the real source for that loss function which is widely used. Keras in their website docs have shown how to create this custom contrastive loss function ( while tensorflow uing keras as underlying api ) have implemented it. 
Was just a suggestion to add it as while creating similarity model it will be easier for training our model with this implemented.

**Comment by darsnack** [comment]
I guess in TF it is part of the "add-ons": https://www.tensorflow.org/addons/api_docs/python/tfa/losses/contrastive_loss

**Comment by arcAman07** [comment]
Yeah , was just a suggestion but I don't think it has to be necessarily be added in this library

**Comment by ToucheSir** [comment]
Given the age and importance of the contrastive loss paper, I don't see why we couldn't implement it. Note that a function for it has been [triaged](https://github.com/pytorch/pytorch/issues/34879) and has a [PR](https://github.com/pytorch/pytorch/pull/54455) for inclusion into PyTorch.

**Comment by arcAman07** [comment]
If the consensus is that this loss function should be added, I will be more than willing to work on this issue.

**Comment by CarloLucibello** [comment]
Let's add it!

**Comment by arcAman07** [comment]
Great will start working on the implementation part

**Comment by CarloLucibello** [comment]
close by #1892
